Signal/src/ViewControllers/PhotoLibrary/PhotoLibrary.swift

```diff
@@ -0,0 +1,288 @@
+                let exportURL = URL(fileURLWithPath: exportPath)
+                exportSession.outputURL = exportURL
+
+                Logger.debug("starting video export")
+                exportSession.exportAsynchronously {
+                    Logger.debug("Completed video export")
+
+                    guard let dataSource = DataSourcePath.dataSource(with: exportURL, shouldDeleteOnDeallocation: true) else {
+                        resolver.reject(PhotoLibraryError.assertionError(description: "Failed to build data source for exported video URL"))
+                        return
+                    }
+
+                    resolver.fulfill((dataSource: dataSource, dataUTI: kUTTypeMPEG4 as String))
+                }
+            }
+        }
+    }
+
+    func outgoingAttachment(for asset: PHAsset) -> Promise<SignalAttachment> {
+        switch asset.mediaType {
+        case .image:
+            return requestImageDataSource(for: asset).map { (dataSource: DataSource, dataUTI: String) in
+                return SignalAttachment.attachment(dataSource: dataSource, dataUTI: dataUTI, imageQuality: .medium)
+            }
+        case .video:
+            return requestVideoDataSource(for: asset).map { (dataSource: DataSource, dataUTI: String) in
+                return SignalAttachment.attachment(dataSource: dataSource, dataUTI: dataUTI)
+            }
+        default:
+            return Promise(error: PhotoLibraryError.unsupportedMediaType)
+        }
+    }
+}
+
+class PhotoCollection {
+    private let collection: PHAssetCollection
+
+    init(collection: PHAssetCollection) {
+        self.collection = collection
+    }
+
+    func localizedTitle() -> String {
+        guard let localizedTitle = collection.localizedTitle?.stripped,
+            localizedTitle.count > 0 else {
+            return NSLocalizedString("PHOTO_PICKER_UNNAMED_COLLECTION", comment: "label for system photo collections which have no name.")
+        }
+        return localizedTitle
+    }
+
+    func contents() -> PhotoCollectionContents {
+        let options = PHFetchOptions()
+        options.sortDescriptors = [NSSortDescriptor(key: "creationDate", ascending: true)]
+        let fetchResult = PHAsset.fetchAssets(in: collection, options: options)
+
+        return PhotoCollectionContents(fetchResult: fetchResult, localizedTitle: localizedTitle())
+    }
+}
+
+class PhotoCollections {
+    let collections: [PhotoCollection]
+
+    init(collections: [PhotoCollection]) {
+        self.collections = collections
+    }
+
+    var count: Int {
+        return collections.count
+    }
+
+    func collection(at index: Int) -> PhotoCollection {
+        return collections[index]
+    }
+}
+
+class PhotoLibrary: NSObject, PHPhotoLibraryChangeObserver {
+    final class WeakDelegate {
+        weak var delegate: PhotoLibraryDelegate?
+        init(_ value: PhotoLibraryDelegate) {
+            delegate = value
+        }
+    }
+    var delegates = [WeakDelegate]()
```

Comment:
TIOLI there's an existing Weak struct if you want to use it. I think it'd be like:

```
var delegates = [Weak<PhotoLibraryDelegate>]
```

Reply:
Done.